File: java/p2/src/p2/types/basic/Schema.java
package p2.types.basic;

import java.util.ArrayList;
import java.util.Hashtable;
import java.util.Iterator;
import java.util.List;

import p2.lang.plan.DontCare;
import p2.lang.plan.Variable;
import p2.types.table.Key;
import p2.types.table.TableName;


public class Schema {
	
	protected TableName name;
	
	protected Hashtable<String, Variable> variables;

	public Schema() {
		this.name = null;
		this.variables = new Hashtable<String, Variable>();
	}

	public Schema(TableName name) {
		this.name = name;
		this.variables = new Hashtable<String, Variable>();
	}
	
	public Schema(TableName name, List<Variable> variables) {
		this.name = name;
		this.variables = new Hashtable<String, Variable>();
		for (Variable variable : variables) {
			this.variables.put(variable.name(), variable);
		}
	}
	
	private Schema(Schema schema) {
		this.name = schema.name;
		this.variables = (Hashtable) schema.variables.clone();
	}
	
	public Schema clone() {
		return new Schema(this);
	}
	
	public TableName name() {
		return this.name;
	}
	
	public int size() {
		return this.variables.size();
	}
	
	public void append(Variable variable) {
		variable.position(this.variables.size());
		this.variables.put(variable.name(), variable);
	}
	
	public boolean contains(Variable variable) {
		return this.variables.containsKey(variable.name());
	}
	
	@Override
	public String toString() {
		return variables().toString();
	}
	
	
	public final List<Class> types() {
		List<Class> types = new ArrayList<Class>();
		for (int position = 0; position < this.variables.size(); position++) {
			for (Variable variable : this.variables.values()) {
				if (variable.position() == position) {
					types.add(variable.type());
					break;
				}
			}
		}
		return types;
	}
	
	public final List<Variable> variables() {
		List<Variable> variables = new ArrayList<Variable>();
		for (int position = 0; position < this.variables.size(); position++) {
			for (Variable variable : this.variables.values()) {
				if (variable.position() == position) {
					variables.add(variable.clone());
					break;
				}
			}
		}
		return variables;
	}
	
	public final Variable variable(String name) {
		return this.variables.get(name).clone();
	}
	
	public final Class type(String name) {
		return this.variables.get(name).type();
	}
	
	public final int position(String name) {
		return this.variables.containsKey(name) ?
				this.variables.get(name).position() : -1;
	}
	
	public final Schema join(Schema inner) {
		Schema join = new Schema();
		for (Variable variable : this.variables()) {
			if (!(variable instanceof DontCare)) {
				join.append(variable);
			}
		}
		
		for (Variable variable : inner.variables()) {
			if (!(variable instanceof DontCare) && !join.contains(variable)) {
				join.append(variable);
			}
		}
		return join;
	}

}

File: lincoln/ruby/lib/types/basic/schema.rb
require "lib/types/table/table_name"
require "lib/lang/plan/variable"
require "rubygems"
class Schema
  def initialize(name, vars)
    @name = name;
    @variable_set = Hash.new
    unless vars.nil?
      vars.each do |v|
        @variable_set[v.name] = v
      end 
    end
  end
    
  attr_accessor :variable_set
    
  def clone
    s = super
    s.variable_set = @variable_set.clone
    return s
  end
  
  def set_schema(schema)
    @name = schema.name
    @variable_set = schema.variables.clone
  end

  def ==(s)
    return (@name == s.name && variables == s.variables)
  end
  
  attr_reader :name

  def size
    @variable_set.length
  end

  def <<(v)
    @variable_set[v.name] = v 
  end

  def contains(v)
    not @variable_set[v.name].nil?
  end

  def to_s
     out = ""
#    out = "("
    i = 0
    @variable_set.sort{|a,b| a[1].position<=>b[1].position}.each do |a|
      out << "," if i > 0
      i += 1
      out << a[1].to_s
    end
 #   out[out.length-1] = ")"     if out.length > 1
#    out << ")"       if out.length == 1
    out
  end

  def types
    out = Array.new
    @variable_set.sort{|a,b| a[1].position<=>b[1].position}.each do |a|
      out << a[1].expr_type if a[1].position >= 0 
    end
    out
  end
  
  def variables
    out = Array.new
    @variable_set.sort{|a,b| a[1].position<=>b[1].position}.each do |a|
      out << a[1].clone if a[1].position >= 0 
    end
    out
  end

  def variable(name)
    @variable_set[name]
  end

  def schema_type(name)
    @variable_set[name].expr_type
  end

  def position(name)
    return (v=variable(name)).nil? ? nil : v.position
  end

  def join(inner)
    joined = Schema.new(nil, nil)
    self.variables.each do |v|
      joined << v
    end

    inner.variables.each do |v|
      unless joined.contains(v)
        joined << v
      end
    end
    joined
  end
end
